### PR TITLE
Make the subscription registry lazy

### PR DESCRIPTION
**Pull request by rullzer**
This will allow to do lazy registration here which should allow for
loading less (or at least only when needed!).

Todo:
- [ ] Extend tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

File: lib/private/Support/Subscription/Registry.php [hunk 1]
```diff
@@ -38,11 +40,35 @@ class Registry implements IRegistry {
...
+	}
+
+	private function getSubscription(): ?ISubscription {
+		if ($this->subscription !== null) {
+			return $this->subscription;
+		}
+
+		if ($this->subscriptionService !== null) {
+			try {
+				$this->subscription = $this->container->query($this->subscriptionService);
```
Comment (kesselb) [inline lib/private/Support/Subscription/Registry.php:64]: The `getSubscription` method looks quite complex ;)

```
	private function getSubscription(): ?ISubscription {
		if ($this->subscription === null && $this->subscriptionService !== null) {
			try {
				$this->subscription = $this->container->query($this->subscriptionService);
			} catch (QueryException $e) {
				// nothing
			}
		}

		return $this->subscription;
	}
```

Would I suggest.

**Review by kesselb — APPROVED** [review]
LGTM :+1:

File: lib/private/Support/Subscription/Registry.php [hunk 2]
```diff
@@ -55,20 +81,29 @@ public function __construct(IConfig $config) {
...
 	public function register(ISubscription $subscription): void {
-		if ($this->subscription !== null) {
+		if ($this->subscription !== null || $this->subscriptionService !== null) {
 			throw new AlreadyRegisteredException();
 		}
 		$this->subscription = $subscription;
 	}
 
+	public function registerService(string $subscriptionService): void {
+		if ($this->subscriptionService !== null || $this->subscriptionService !== null) {
```
Comment (kesselb) [inline lib/private/Support/Subscription/Registry.php:91]:
```suggestion
		if ($this->subscriptionService !== null) {
```

Comment (kesselb) [reply lib/private/Support/Subscription/Registry.php:91]: Or 

```suggestion
		if ($this->subscription !== null || $this->subscriptionService !== null) {
```

:man_shrugging:

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

File: tests/lib/Support/Subscription/DummySubscription.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Test\Support\Subscription;
+
+
```
Comment (MorrisJobke) [inline tests/lib/Support/Subscription/DummySubscription.php:27]: Drop this line and CS is also fine with it.

Comment (MorrisJobke) [reply tests/lib/Support/Subscription/DummySubscription.php]: Done.

**Comment by ChristophWurst** [comment]
Could we deprecate the other registration method?

**Comment by rullzer** [comment]
> Could we deprecate the other registration method?

https://github.com/nextcloud/server/pull/21759/files#diff-ee1ee14a4eccf65986921c58fcc12eb4R45

**Review by ChristophWurst — APPROVED** [review]

File: lib/private/Support/Subscription/Registry.php [hunk 4]
```diff
@@ -55,20 +76,29 @@ public function __construct(IConfig $config) {
 	 * @since 17.0.0
```
Comment (ChristophWurst) [inline lib/private/Support/Subscription/Registry.php:76]: I looked here :wink:

**Comment by MorrisJobke** [comment]
Ready for merge or not?

**Comment by MorrisJobke** [comment]
@ChristophWurst Shouldn't we move this as well to the `IRegistrationContext`? so it can be done in the `boot()` method?

**Pull request merged**